### PR TITLE
Use last successful fit as input to subsequent fits in EngDiff sequential fitting

### LINKED ISSUE #31527: EngDiff UI sequential fitting should only updates parameters if fit successful
Sequential fit should continue if the fit didn’t converge using the last good fit parameters (not a problem with BackToBackExponential but is an issue with unmodified BackToBackConvPV)

### PR DESCRIPTION
**Pull request by RichardWaiteSTFC**
**Description of work.**

Sequential fitting now only updates the input to the next fit with the last successful fit (i.e. it checks the output status of the Fit algorithm). If a fit is unsuccessful the parameters from the last iteration of Fit algorithm are still returned, but now the fit status is also output in these tables to make it clear to the user which fits succeeded without them having to check the log (mantid labels a fit a failure if it converges on parameter step size rather than change in chisq - but sometimes you will want to accept these fits).

Have added a unit test to check that the output of an unsuccessful fit is not passed to the subsequent fit as input (this is hard to test manually but should be obvious from code review).

**To test:**
(1) Follow instructions in #29726 (note in step 11 you are no longer prompted that the function will be overwritten)
(2) Check that there is an additional column in the `_model` workspace in the `_fit` group of workspaces that is the status of the fit

Fixes #31527 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by DannyHindson — COMMENTED** [review]
I've tested this and it works fine. I noticed one possible glitch but it looks like it may relate to the fit property browser rather than any of your eng UI code.
If one of the fits fails you can end up with this situation in the fit property browser where the top summary says success but on the "status" line below it provides a failure message ("Changes in parameter value..."):

image

Since this is trying to show the result of multiple\sequential fits I guess it's never going to be 100% clear but ideally they'd both show failure.
Not sure how this can happen. Perhaps in FitPropertyBrowser::finishHandle something around lines 1645-1649 fails (eg can't retrieve the OutputChi2overDoF parameter) when the fit has failed meaning the code never emits the changeWindowTitle signal.

If you agree this is separate\unrelated then I'll approve

File: scripts/Engineering/gui/engineering_diffraction/tabs/fitting/plotting/plot_view.py [hunk 1]
```diff
@@ -148,7 +148,7 @@ def read_fitprop_from_browser(self):
     def update_browser(self, status, func_str, setup_name):
         self.fit_browser.fitResultsChanged.emit(status)
```
Comment (RichardWaiteSTFC) [inline scripts/Engineering/gui/engineering_diffraction/tabs/fitting/plotting/plot_view.py:149]: This is called after each iteration of the sequential fit and updates the status displayed in the browser. So at the end of a sequential fit it should display the status of the last fit. I did wonder whether to just update the status after the last fit only - I can do that in this PR by separating the updating of the status from the loading/saving function as setup.

Comment (DannyHindson) [reply scripts/Engineering/gui/engineering_diffraction/tabs/fitting/plotting/plot_view.py:149]: Thanks - that explains it. I don't think it matters whether you call fitResultsChanged more or less frequently. The point is that the window title isn't updated in sync with this so the title can say "success" while the status says something which is not success.

Have a look in FitPropertyBrowser::finishHandle. This emits both fitResultsChanged and changeWindowTitle. So maybe you need to emit changeWindowTitle also in the engineering code?

Comment (RichardWaiteSTFC) [reply scripts/Engineering/gui/engineering_diffraction/tabs/fitting/plotting/plot_view.py:149]: Ah I understand now, I will try emitting the `changeWindowTitle` - thanks for the suggestion!

**Comment by DannyHindson** [comment]
Thanks for the change. So the two status indicators in the fit property browser now line up which is good. I noticed that it's actually reporting the status of last fit to complete. So you can have two successes and one failure and both statuses will say "success" if one of the successful ones completes last. I think we should be indicating at least one of the fits failed. I don't know what is feasible here - I guess we could just show "failure" or if we can say something like "x succeeded, y failed" that would be even better.

As I mentioned earlier though - this doesn't relate to the changes you've made in this PR so we could deal with separately

**Comment by RichardWaiteSTFC** [comment]
> Thanks for the change. So the two status indicators in the fit property browser now line up which is good. I noticed that it's actually reporting the status of last fit to complete. So you can have two successes and one failure and both statuses will say "success" if one of the successful ones completes last. I think we should be indicating at least one of the fits failed. I don't know what is feasible here - I guess we could just show "failure" or if we can say something like "x succeeded, y failed" that would be even better.
> 
> As I mentioned earlier though - this doesn't relate to the changes you've made in this PR so we could deal with separately

I had only intended to show the status of the last fit but I like the idea of printing some sort of summary status for the sequential fit as a whole. However, it is difficult to summarise the combined status in the browser title etc. All the details relevant to the user (including the status of each fit) is stored in a table workspace so maybe this is sufficient for now? I'm open to any ideas though, but if it's OK I'd like to address them in a separate PR related to issue #31688.

**Comment by RichardWaiteSTFC** [comment]
> I had only intended to show the status of the last fit (hence my comment above about updating the status too often - it's currently done after each run) but I like the idea of printing some sort of summary status for the sequential fit as a whole. However, it is difficult to summarise the combined status in the browser title etc. 

So in a PR related to issue #31688 I am think maybe in the window title I report something like `m fits out of n successful` and in the fitbrowser status I keep the status of the last fit. Does that sound good?

**Comment by DannyHindson** [comment]
It's fine to cover in #31688 so I'll approve this PR.

I think if we had complete flexibility over what to show then all\both status messages would show some kind of summary of all the runs fitted - not sure showing "success" because the last one worked (even though several other ones could have failed) adds a lot of value. I appreciate though that some of this is probably limited by the code in the core fit property browser though so if this restricts us to just showing the last result in one of the two status lines then we'll have to live with that.

**Review by DannyHindson — DISMISSED** [review]
Looks good.

I'll rerun the RHEL job now to see if it squeezes through. As discussed this has failed due to a known issue with the ISIS_WISHSingleCrystalReduction test that sometimes gets skipped if not enough memory

**Comment by gemmaguest** [comment]
@RichardWaiteSTFC Does this need release notes?

**Comment by RichardWaiteSTFC** [comment]
> @RichardWaiteSTFC Does this need release notes?

I will add some now - will need to merge in master as this branch doesn't have 6.2 release notes!

**Review by DannyHindson — APPROVED** [review]

**Pull request merged**